ir/src/boundary_stmts.rs

```diff
@@ -0,0 +1,175 @@
+    /// - A boundary constraint has already been declared for the specified column and boundary.
+    pub(super) fn insert(
+        &mut self,
+        symbol_table: &mut SymbolTable,
+        stmt: &BoundaryStmt,
+    ) -> Result<(), SemanticError> {
+        match stmt {
+            BoundaryStmt::Variable(_) => unimplemented!(),
+            BoundaryStmt::Constraint(constraint) => {
+                // validate the expression
+                let expr = constraint.value();
+                validate_expression(symbol_table, &expr)?;
+
+                // add the constraint to the specified boundary for the specified trace
+                let col_type = symbol_table.get_type(constraint.column())?;
+                let result = match col_type {
+                    IdentifierType::TraceColumn(column) => match column.trace_segment() {
+                        0 => match constraint.boundary() {
+                            ast::Boundary::First => self.boundary_constraints[0]
+                                .0
+                                .insert(column.col_idx(), expr),
+                            ast::Boundary::Last => self.boundary_constraints[0]
+                                .1
+                                .insert(column.col_idx(), expr),
+                        },
+                        1 => match constraint.boundary() {
+                            ast::Boundary::First => self.boundary_constraints[1]
+                                .0
+                                .insert(column.col_idx(), expr),
+                            ast::Boundary::Last => self.boundary_constraints[1]
+                                .1
+                                .insert(column.col_idx(), expr),
+                        },
+                        _ => unimplemented!(),
+                    },
+                    _ => {
+                        return Err(SemanticError::InvalidUsage(format!(
+                            "Identifier {} was declared as a {}, not as a trace column",
+                            constraint.column(),
+                            col_type
+                        )));
+                    }
+                };
+
+                // raise an error if multiple constraints were applied to the same boundary
+                if result.is_some() {
+                    return Err(SemanticError::TooManyConstraints(format!(
+                        "A boundary constraint was already defined for {} '{}' at the {}",
+                        col_type,
+                        constraint.column(),
+                        constraint.boundary()
+                    )));
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Recursively validates the BoundaryExpression.
+///
+/// # Errors
+/// Returns an error if the expression includes a reference to a public input that hasn't been
+/// declared or to an invalid index in an existing public input.
+/// TODO: Complete implementation of validation.
```

Comment:
What's missing in the implementation of the validation? Is there an issue tracking this?

Reply:
That was for me to check before submitting this PR. I think all the cases should be covered now.